Preprocessing
================

  - [Reading count matrix and Initialize SCE
    object](#reading-count-matrix-and-initialize-sce-object)
      - [Get All location by reading
        folders](#get-all-location-by-reading-folders)
  - [QC](#qc)
      - [Calculate QC metrics and filter
        cells](#calculate-qc-metrics-and-filter-cells)
      - [Create filtered expression
        object](#create-filtered-expression-object)
  - [Overview of the dataset](#overview-of-the-dataset)
      - [Plotting markers](#plotting-markers)
  - [Processing glial dataset](#processing-glial-dataset)
      - [pagoda2](#pagoda2)
      - [correcting count matrix with
        scde](#correcting-count-matrix-with-scde)
      - [mapping science annotations](#mapping-science-annotations)
      - [SCENIC](#scenic)
      - [Saving as SingleCellExperiment for
        scanpy](#saving-as-singlecellexperiment-for-scanpy)
      - [Generating pagoda2 web app](#generating-pagoda2-web-app)

This notebook describe the full processing pipeline prior to trajectory
analysis

``` r
source("../_helpers/process.helpers.R")
source("../_helpers/scde.helpers.R")
```

# Reading count matrix and Initialize SCE object

## Get All location by reading folders

``` r
library(SingleCellExperiment)
raw_counts=as.matrix(read.csv("SS2/output/raw_counts.csv",row.names = 1))
anno = read.csv("SS2/output/anno.csv",row.names = 1)

reads <- SingleCellExperiment(
  assays = list(counts = raw_counts),
  colData = anno
)

library(stringr)
time = factor(as.numeric(str_extract(reads$timepoints, "\\-*\\d+\\.*\\d*")))
names(time) = colnames(reads)
reads$time=time
```

# QC

## Calculate QC metrics and filter cells

``` r
library(scater)
library(grid)
library(egg)
reads <- calculateQCMetrics(reads, feature_controls = list(ERCC = 1:length(erccnames)))

reads = reads[,reads$total_features_by_counts!=0]

q1=ggplot() + 
  geom_histogram(aes(reads$total_counts,fill=reads$exp),bins = 100) + 
  geom_vline(xintercept = 0.5e5, color = "red", size=1) +
  xlab("total counts")

filter_by_total_counts <- (reads$total_counts > 0.5e5)

q2 = ggplot() + 
  geom_histogram(aes(reads$total_features_by_counts,fill=reads$exp),bins = 100) + 
  geom_vline(xintercept = 1000, color = "red", size=1) +
  xlab("total features")

filter_by_expr_features <- (reads$total_features_by_counts > 1000)

q3 = plotColData(reads,x="total_features_by_counts",y="pct_counts_ERCC", colour_by = "exp") + 
  geom_hline(yintercept = 25, color = "red", size=1) +theme(legend.position = "bottom")

filter_by_ERCC <- reads$pct_counts_ERCC < 25

reads$use <- (
  # sufficient features (genes)
  filter_by_expr_features &
    # sufficient molecules counted
    filter_by_total_counts &
    # sufficient endogenous RNA
    filter_by_ERCC
)
```

## Create filtered expression object

``` r
reads.filtered = reads[-(1:length(erccnames)),reads$use]
tps=as.character(reads.filtered$timepoints)

reads.filtered$timepoints=factor(gsub("-","",gsub("[.]","",gsub("\\d", "", tps))))

names(reads.filtered$timepoints)=colnames(reads.filtered)
names(reads.filtered$time)=colnames(reads.filtered)
names(reads.filtered$exp)=colnames(reads.filtered)

saveRDS(reads.filtered,"output/reads.filtered.rds")
```

# Overview of the dataset

``` r
library(Seurat)

cnts=counts(reads.filtered)[apply(counts(reads.filtered),1,function(x) sum(x>0))!=0,]
p2=p2.wrapper(cnts)

p2$plotEmbedding(type = "PCA",embeddingType = "UMAP",clusterType = "leiden",mark.clusters=T)
save(p2,file="output/p2_Overview.RData")
```

## Plotting markers

``` r
png("allcells.png",width = 15,height = 5,units = "in",res=600);par(mfrow=c(1,3));p2_all$plotEmbedding("PCA","UMAP",mark.clusters = T);p2_all$plotEmbedding("PCA","UMAP",colors=p2_all$counts[,"Sox10"]);p2_all$plotEmbedding("PCA","UMAP",groups = sel);dev.off()


genes=c("Isl1","Pecam1","Kdr",
        "Olig3","Prrx1","Prrx2",
        "Sox2","Msx1","Cntn2",
        "Star","Fdxr","Hsd3b1",
        "Aif1","Vav1","Ermap",
        "Sp7","Epcam","Krt8","Krt18")

for (gene in genes){
  png(paste0(gene,".png"),width = 5,height = 5,res = 600,units = "in")
  p2_all$plotEmbedding("PCA","UMAP",colors=p2_all$counts[,gene])
  title(gene)
  dev.off()
}

png("Sox10.png",width = 5,height = 5,res = 600,units = "in")
p2_all$plotEmbedding("PCA","UMAP",colors=p2_all$counts[,gene])
title("Sox10")
dev.off()


png("selection.png",width = 5,height = 5,res = 600,units = "in")
p2_all$plotEmbedding("PCA","UMAP",groups=sel)
title("selection")
dev.off()

png("leiden_clustering.png",width = 5,height = 5,res = 600,units = "in")
p2_all$plotEmbedding("PCA","UMAP",)
title("leiden")
dev.off()
```

# Processing glial dataset

## pagoda2

``` r
res=p2$getDifferentialGenes("PCA","leiden",upregulated.only = T)

select=rownames(p2$counts)[p2$clusters$PCA$leiden%in%c(7,5,23,16,15,9,12,3,6,13,11,17)]


cnts_Sox10=Matrix::t(p2$misc$rawCounts[select,])
cnts_Sox10=cnts_Sox10[apply(cnts_Sox10,1,function(x) sum(x>0))!=0,]


p2=p2.wrapper(cnts_Sox10)

pal=doPalantir(p2$reductions$PCA,n_neighbors = 100,min_dist = .3,diff_knn = 80,n_eig = 10)


tm=as.character(reads.filtered$time)
tm[is.na(tm)]="Adult"
tm[tm=="0"]="P0"
tm[tm=="2"]="P2"
tm[tm=="6"]="P6"
names(tm)=names(reads.filtered$time)

cells_P6=names(tm[tm=="P6"])

tm[cells_P6[!grepl("([A-Q][1-9]|[A-Q][0-1][0-2])_un",cells_P6)]]="P10"

tm=factor(tm,levels=c("9.5","10.5","11.5","12.5","13.5","14.5","16.5","18.5","P0","P2","P6","P10","Adult"))

cc = scales::seq_gradient_pal("yellow","darkblue")(seq(0,1,length.out=13))


p2$embeddings$Diff$UMAP=pal$umap
p2$reductions$Diff=pal$ms_data

p2$clusters$labels$devtime=tm[rownames(p2$counts)]
p2$clusters$labels$location=reads.filtered$timepoints[rownames(p2$counts)]

library(CytoTRACE)
results <- CytoTRACE(t(as.matrix(p2$misc$rawCounts)),ncores = 4)

write.csv(as.matrix(p2$counts),"scenic/resources/expr.csv")
```

## correcting count matrix with scde

``` r
counts=as.matrix(t(p2$misc$rawCounts))
mode(counts)<-"integer"
cdb <- gsub(":.*","",colnames(counts));
n.cores=10
min.cell.genes <- 3e3;min.cell.reads <- 1e3;min.gene.reads <- 10;
min.gene.cells <- 5;min.nonfailed <- 8;n.groups <- 10;trim <- 3;
res <- scde.process.dataset(counts,"sensory",batch=cdb,skip.pca = T)

fpm <- log10(exp(scde::scde.expression.magnitude(res$knn, res$cd))+1)

write.csv(fpm,"output/fpm.csv")
```

## mapping science annotations

``` r
par(mfrow=c(3,2))
tsne=read.table("http://pklab.med.harvard.edu/ruslan/neural_crest/tSNE_main_Fig1.txt",
                header = F,sep=" ",skip=1,row.names = 1,comment.char = "")

emb=p2$embeddings$SCENIC$dUMAP
rownames(emb)=sapply(rownames(p2$counts),function(x) stringr::str_replace(strsplit(x,"_unique")[[1]][1],":","_"))
emb=data.frame(emb)
load("~/Projects/Glial/SS2/output/crest.rda")
emb$col="lightgrey"
tsne$V4=crest$clcol
emb[rownames(tsne),]$col=tsne$V4



pl1=ggplot()+geom_point(aes(tsne$V2,tsne$V3),color=tsne$V4,size=3)+theme_void()


emb1=emb[emb$col=="lightgrey",]
emb2=emb[emb$col!="lightgrey",]
pl2=ggplot()+geom_point(aes(emb1$X0,emb1$X1),color=emb1$col)+
  geom_point(aes(emb2$X0,emb2$X1),color=emb2$col)+theme_void()


k=1
cols=unique(emb[emb$col!="lightgrey","col"])
for (col in cols){
  emb1=emb[emb$col!=col,]
  emb2=emb[emb$col==col,]
  pl2=ggplot()+geom_point(aes(emb1$X0,emb1$X1),color="lightgrey")+
    geom_point(aes(emb2$X0,emb2$X1),color=col)+theme_void()
  ggsave(paste0(c("figures/pl2_",(k),".png"),collapse = ""),pl2,width = 5,height = 5,dpi = 600)
  k=k+1
}



cowplot::plot_grid(pl1,pl2)

tsne=read.table("http://pklab.med.harvard.edu/ruslan/neural_crest/tSNE_Wnt1_FigS5.txt",
                header = F,sep=" ",skip=1,row.names = 1,comment.char = "")

emb$col=NA
emb[rownames(tsne),]$col=tsne$V4

plot(tsne$V2,tsne$V3,col=tsne$V4,pch = 19)
plot(emb$x,emb$y,col=emb$col,pch = 19)


tsne=read.table("http://pklab.med.harvard.edu/ruslan/neural_crest/tSNE_Sox10_FigS5.txt",
                header = F,sep=" ",skip=1,row.names = 1,comment.char = "")

emb$col=NA
emb[rownames(tsne),]$col=tsne$V4

plot(tsne$V2,tsne$V3,col=tsne$V4,pch = 19)
plot(emb$x,emb$y,col=emb$col,pch = 19)
```

## SCENIC

### Obtaining cisTarget databases

``` bash
wget -P  scenic/cisTarget_databases https://resources.aertslab.org/cistarget/motif2tf/motifs-v9-nr.mgi-m0.001-o0.0.tbl \
  https://resources.aertslab.org/cistarget/databases/mus_musculus/mm10/refseq_r80/mc9nr/gene_based/mm10__refseq-r80__500bp_up_and_100bp_down_tss.mc9nr.feather \
  https://resources.aertslab.org/cistarget/databases/mus_musculus/mm10/refseq_r80/mc9nr/gene_based/mm10__refseq-r80__10kb_up_and_down_tss.mc9nr.feather
```

### Running SCENIC

``` bash
docker run -it --rm -v $(pwd)/scenic:/scenicdata \
    aertslab/pyscenic:0.10.0 pyscenic grn \
        --num_workers 20 \
        -o /scenicdata/output/adjacencies.tsv \
        /scenicdata/resources/expr.csv \
        /scenicdata/resources/mm_mgi_tfs.txt

docker run --rm -v $(pwd)/scenic:/scenicdata \
    aertslab/pyscenic:0.10.0 pyscenic ctx \
        /scenicdata/output/adjacencies.tsv \
        /scenicdata/cisTarget_databases/mm10__refseq-r80__10kb_up_and_down_tss.mc9nr.feather \
        /scenicdata/cisTarget_databases/mm10__refseq-r80__500bp_up_and_100bp_down_tss.mc9nr.feather \
        --annotations_fname /scenicdata/cisTarget_databases/motifs-v9-nr.mgi-m0.001-o0.0.tbl \
        --expression_mtx_fname /scenicdata/resources/expr.csv \
        --mode "dask_multiprocessing" \
        --output /scenicdata/output/regulons_mm10.csv \
        --num_workers 20

docker run --rm -v $(pwd)/scenic:/scenicdata \
    aertslab/pyscenic:0.10.0 pyscenic aucell \
    /scenicdata/resources/expr.csv \
    /scenicdata/output/regulons_mm10.csv \
    -o /scenicdata/output/auc_mtx_mm10.csv \
    --num_workers 20
```

## Saving as SingleCellExperiment for scanpy

``` r
write.csv(p2$clusters$labels$devtime,"output/devtime.csv")
write.csv(p2$clusters$labels$location,"output/location.csv")

p2$reductions$aucDiff=read.csv("output/ms_data_auc.csv",row.names = 1)
p2$embeddings$aucDiff$UMAP=read.csv("output/umap_2d.csv",row.names = 1)

adata <- SingleCellExperiment(list(logcounts=t(p2$counts)),
    colData=DataFrame(leiden=p2$clusters$PCA$leiden,
                      devtime=as.character(p2$clusters$labels$devtime),
                      location=as.character(p2$clusters$labels$location),
                      cytoTRACE=results$CytoTRACE,
                      cytoTRACErank=results$CytoTRACErank),
    metadata=list(study="glia")
)



reducedDims(adata) <- list(PCA=p2$reductions$PCA,
                         UMAP=p2$embeddings$PCA$UMAP)


save(adata,file = "output/adata.RData")
```

## Generating pagoda2 web app

``` r
## add cluster selection

p2$n.cores=10
go.env <- p2.generate.mouse.go(p2)
p2$testPathwayOverdispersion(setenv = go.env,
                             recalculate.pca=F,
                             correlation.distance.threshold = 0.95,verbose=T)

myGeneNames <- colnames(p2$counts)
goSets <- p2.generate.mouse.go.web(myGeneNames)
deSets <- get.de.geneset(p2, groups = p2$clusters$PCA$leiden, prefix = 'de_')
geneSets <- c(goSets, deSets)


additionalMetadata <- list()

additionalMetadata$leiden <- p2.metadata.from.factor(p2$clusters$PCA$leiden, displayname = 'Leiden', s = 0.7, v = 0.8,start = 0, end = 0.5,pal = tableau_color_pal(palette = "Tableau 20")(nlevels(p2$clusters$PCA$leiden)))


cc = scales::seq_gradient_pal("yellow","darkblue")(seq(0,1,length.out=nlevels(p2$clusters$labels$devtime)))

additionalMetadata$time <- p2.metadata.from.factor(p2$clusters$labels$devtime, 
                                                   displayname = 'Time', 
                                                   s = 0.7, 
                                                   v = 0.8,
                                                   pal=cc)

additionalMetadata$tp <- p2.metadata.from.factor(p2$clusters$labels$location, 
                                                 displayname = 'Location', 
                                                 s = 0.7, 
                                                 v = 0.8,
                                                 start = 0, 
                                                 end = 0.5)


p2$makeGeneKnnGraph(n.cores=20)

p2w <- make.p2.app(
  p2,
  dendrogramCellGroups = p2$clusters$PCA$leiden,
  additionalMetadata = additionalMetadata,
  geneSets = geneSets,
  show.clusters = FALSE # Hide the clusters that were used for the dendrogram from the metadata
)

p2w$serializeToStaticFast("p2w_Glia.bin")
save(p2w,file="p2w_Glia.RData")
```
